htlcswitch/link.go

```diff
@@ -2830,31 +2830,46 @@ func (l *channelLink) processExitHop(pd *lnwallet.PaymentDescriptor,
 		HtlcID: pd.HtlcIndex,
 	}
 
-	event, _, err := l.cfg.Registry.NotifyExitHopHtlc(
-		invoiceHash, pd.Amount, pd.Timeout, int32(heightNow),
-		circuitKey, l.hodlQueue.ChanIn(), payload,
-	)
-
-	switch err {
-
-	// Cancel htlc if we don't have an invoice for it.
-	case channeldb.ErrInvoiceNotFound:
+	// sendFailure is a closure which sends a htlc failure with a custom
+	// failure detail and returns a nil error.
+	sendFailure := func(detail FailureDetail) error {
 		failure := &ForwardingError{
-			FailureMessage: lnwire.NewFailIncorrectDetails(pd.Amount, heightNow),
-			FailureDetail:  &FailureDetailInvoiceNotFound{paymentHash: invoiceHash},
+			FailureMessage: lnwire.NewFailIncorrectDetails(
+				pd.Amount, heightNow,
+			),
+			FailureDetail: detail,
 		}
-		l.sendHTLCError(pd.HtlcIndex, failure, obfuscator, pd.SourceRef)
 
+		l.sendHTLCError(pd.HtlcIndex, failure, obfuscator, pd.SourceRef)
 		return nil
+	}
 
-	// No error.
-	case nil:
+	event, result, err := l.cfg.Registry.NotifyExitHopHtlc(
+		invoiceHash, pd.Amount, pd.Timeout, int32(heightNow),
+		circuitKey, l.hodlQueue.ChanIn(), payload,
+	)
 
-	// Pass error to caller.
-	default:
+	if err == channeldb.ErrInvoiceNotFound {
+		return sendFailure(
+			&FailureDetailInvoiceNotFound{paymentHash: invoiceHash},
+		)
+	} else if err != nil {
 		return err
 	}
 
+	// If the htlc was not suitable for paying our invoice, fail it.
+	// This allows for failures with a specific detail attached, rather than
+	// falling through to processHodlEvent and treating all failures like hodl
+	// cancels.
+	switch result {
+	case invoices.ResultAmountTooLow:
+		return sendFailure(&FailureDetailInvoiceUnderpaid{})
+	case invoices.ResultExpiryTooSoon:
+		return sendFailure(&FailureDetailInvoiceExpiryTooSoon{})
+	case invoices.ResultInvoiceAlreadyCanceled:
+		return sendFailure(&FailureDetailInvoiceAlreadyCancelled{})
```

Comment:
What about the other results?

It also make me wonder whether we should return `FailureDetail` objects instead of `UpdateResult`s from the registry.

Reply:
The other results don't need `htlcNotifier` events, replays will already have been notified and settles are handled by `processHodlEvent` when we settle the invoice (and needs to be there because we call it in multiple places).